src/validation.cpp

```diff
@@ -1060,28 +1184,92 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+    // Check static package limits before taking the mempool lock.
+    const unsigned int package_count = txns.size();
+    if (package_count > MAX_PACKAGE_COUNT) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-many-transactions");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    const int64_t total_size = std::accumulate(txns.cbegin(), txns.cend(), 0,
+        [](int64_t sum, const auto& tx) { return sum + GetVirtualTransactionSize(*tx); });
+    if (total_size > MAX_PACKAGE_SIZE * 1000) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-large");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(package_count);
+    std::transform(txns.cbegin(), txns.cend(), std::back_inserter(workspaces), [](const auto& tx) {
+        return Workspace(tx);
+    });
+
+    LOCK(m_pool.cs);
+
+    std::map<const uint256, const MempoolAcceptResult> results;
+    // Do all PreChecks first and fail fast to avoid running expensive script checks when unnecessary.
+    for (Workspace& ws : workspaces) {
+        if (!PreChecks(args, ws)) {
+            package_state.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+            // Exit early to avoid doing pointless work. Update the failed tx result; the rest are unfinished.
+            results.emplace(ws.m_ptx->GetHash(), MempoolAcceptResult::Failure(ws.m_state));
+            return PackageMempoolAcceptResult(package_state, std::move(results));
+        }
+        m_view.PackageAddTransaction(ws.m_ptx);
+    }
+
+    // We have now verified all inputs are available and there are no conflicts in the package by
+    // marking coins (temporarily) spent in m_view. However, since script checks will assert that
+    // each coin has not been spent as a sanity check, we need to clear m_view.m_temp_spent so that
+    // script checks don't fail on those asserts.
+    m_view.ClearTemporarySpends();
+
+    for (Workspace& ws : workspaces) {
+        PrecomputedTransactionData txdata;
+        if (!PolicyScriptChecks(args, ws, txdata)) {
+            // Exit early to avoid doing pointless work. Update the failed tx result; the rest are unfinished.
+            package_state.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+            results.emplace(ws.m_ptx->GetHash(), MempoolAcceptResult::Failure(ws.m_state));
+            return PackageMempoolAcceptResult(package_state, std::move(results));
+        }
+        // When test_accept=true, transactions that pass PolicyScriptChecks are valid because there are
+        // no further mempool checks (passing PolicyScriptChecks implies passing ConsensusScriptChecks).
+        results.emplace(ws.m_ptx->GetHash(),
+                        MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees));
+    }
+
+    m_view.ClearTemporaryCoins();
```

Comment:
Is this needed?

Reply:
![image](https://user-images.githubusercontent.com/25183001/112912067-8760af80-90ab-11eb-9967-97a6c6b3442d.png)

(removed for now)